File: PROBLEMS/arrays/pigeonhole_query_corrected.py
def main():
    N, Q = map(int, input().split())
    pigeons = list(range(N + 1))  # array that says for each pigeon (position) which nest it is (value)

    for _ in range(Q):
        query_type = int(input())

        if query_type == 1:
            P, H = map(int, input().split())
            pigeons[P] = H

        elif query_type == 2:
            nest_count = {}  # Mapeia cada ninho para a quantidade de pombos
            
            for i in range(1, N + 1):  # Contar pombos em cada ninho
                nest_count[pigeons[i]] = nest_count.get(pigeons[i], 0) + 1

            n_nests = sum(1 for count in nest_count.values() if count > 1)  # Verificar se o ninho tem mais de um pombo
            print(n_nests)

if __name__ == "__main__":
    main()

File: PROBLEMS/arrays/pigeonhole_query_corrected.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int N;
    int Q;
    cin >> N >> Q;
    vector <int> pigeons(N + 1); // array that say for each pigeon (position) which nest he is (value)
    
    for (int i = 0; i < N + 1; i++) {
        pigeons[i] = i;
    }

    for (int i = 0; i < Q; i++) {
        int query_type;
        cin >> query_type;

        if (query_type == 1) {
            int P;
            int H;
            cin >> P >> H;
            pigeons[P] = H;
        } 

        else if (query_type == 2) {
            map<int, int> nest_count; // Mapeia cada ninho para a quantidade de pombos
            
            for (int i = 1; i <= N; i++) { // Contar pombos em cada ninho
                nest_count[pigeons[i]]++;
            }

            int n_nests = 0;
            for (auto& pair : nest_count) {
                if (pair.second > 1) { // Verificar se o ninho tem mais de um pombo
                    n_nests++;
                }
            }
            
            cout << n_nests << endl;
        }
    }
}
